renku/core/commands/login.py

```diff
@@ -0,0 +1,116 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2018-2021 - Swiss Data Science Center (SDSC)
+# A partnership between École Polytechnique Fédérale de Lausanne (EPFL) and
+# Eidgenössische Technische Hochschule Zürich (ETHZ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Logging in to a Renku deployment."""
+
+import os
+import sys
+import time
+import urllib
+import uuid
+import webbrowser
+from pathlib import posixpath
+
+import requests
+
+from renku.core import errors
+from renku.core.incubation.command import Command
+from renku.core.models.enums import ConfigFilter
+from renku.core.utils import communication
+from renku.core.utils.urls import parse_authentication_endpoint
+
+CONFIG_SECTION = "http"
+
+
+def login_command():
+    """Return a command for logging in to the platform."""
+    return Command().command(_login)
+
+
+def _login(client, endpoint):
+    parsed_endpoint = _parse_endpoint(client, endpoint)
+    query = urllib.parse.urlencode({"cli_token": str(uuid.uuid4())})
+
+    communication.echo(
+        f"Please log in at {parsed_endpoint.geturl()} on your browser.\n"
+        "Once completed, you may close the browser window."
+    )
+
+    login_url = _get_url(parsed_endpoint, "/api/auth/login", query)
+    webbrowser.open_new_tab(login_url)
+
+    info_url = _get_url(parsed_endpoint, "/api/auth/info", query)
```

Comment:
The login URL should really not just show the regular login page, but something like "Application 'Renku Commandline' wants access to your renkulab.io account. Authorize? Yes | No". Otherwise this is a pretty big security risk.

I could just post a link to facebook or something with "Hey, checkout my cool renku project" and it links to the /api/auth/login URL with an UUID that I (the attacker) know, and then I can just monitor the endpoint to get a users token and wait for someone to click the link. Even worse if the user is already logged in, in which case clicking the link is already enough and they don't need to even enter their credentials. As soon as they click the link, they would have basically lost.

With the more traditional approach of having a local webserver as a callback, this is somewhat mitigated as you can limit the callback URL to being `localhost`.

But either way, we need a page that tells the user what is going to happen and where they have to explicitly confirm that this is what they want.